code/code/misc/tweaks.cc

```diff
@@ -0,0 +1,268 @@
+}
+
+tweakTypeT & operator++(tweakTypeT &c, int)
+{
+  return c = (c == MAX_TWEAK_TYPES) ? TWEAK_NONE : tweakTypeT(c+1);
+}
+
+tweakInfoT::~tweakInfoT()
+{
+}
+
+tweakInfoT::tweakInfoT(): 
+    tweaks(MAX_TWEAK_TYPES + 1, nullptr), 
+    loaded(false)
+{
+  loaded=false;
+}
+
+int tweakInfoT::loadTweaks()
+{
+  TDatabase db(DB_SNEEZY);
+
+  //On load fill out all tweaks with 1.0.  This will make adding new tweaks easier by eliminating the need to write a migration.
+  for(tweakTypeT i=TWEAK_NONE;i<MAX_TWEAK_TYPES;i++){
+    tweaks[i] = new tweakEntry(0,1.0, 1.0, 0.0);
+  }
+
+  //the globaltweaks table contains multiple rows per tweak type
+  //We only want to keep the last tweak update of each tweak type in memory, so we do a greatest n per group query
+  db.query("select tweak_id, tweak_type, tweak_value, tweak_target, tweak_rate from globaltweaks a "
+           "where tweak_id = (select max( tweak_id ) from globaltweaks b where a.tweak_type = b.tweak_type)"
+  );
+  //enter in the proper values. Is it possible for the query to fail and the mud still be running?
+  int i = 0;
+  while(db.fetchRow()){
+    tweakTypeT tweak_id=(tweakTypeT) convertTo<int>(db["tweak_type"]);
+    int tid=(int) convertTo<int>(db["tweak_id"]);
+    double val=(double) convertTo<double>(db["tweak_value"]);
+    double tar=(double) convertTo<double>(db["tweak_target"]);
+    double rate=(double) convertTo<double>(db["tweak_rate"]);
+    tweaks[tweak_id]->id = tid;
+    tweaks[tweak_id]->cvalue = val;
+    tweaks[tweak_id]->tvalue = tar;
+    tweaks[tweak_id]->rate = rate;
+    i++;
+  }
+
+  loaded=true;
+  return i;
+}
+
+tweakInfoT tweakInfo;
+
+//recieved after immortal enters 'info tweak <arg>'
+void tweakInfoT::doTweak(TBeing *b, const char *arg)
+{
+  char arg2[256];
+  arg=one_argument(arg, arg2, cElements(arg2));
+
+  if (!*arg2) {  //user just typed info tweak
+    showTweaks(b);
+    return;
+  }
+
+  //determine type of tweak
+  tweakTypeT t;
+  for(t=TWEAK_LOADRATE;t<MAX_TWEAK_TYPES;t++){
+    if (is_abbrev(arg2, getTweakName(t))) {
+      //a tweak type has been found
+      if (!*arg) {
+        showTweaks(b, t);
+        return;
+      }
+      else {
+        handleTweak(b, t, arg);
+        return;
+      }
+    }
+  }
+
+  //if they got here they typed info tweak gibberish
+  showTweaks(b);
+  showTweakOptions(b);
+  return;
+}
+
+//deal with options for specific tweak t
+void tweakInfoT::handleTweak(TBeing *b, tweakTypeT t, const char *arg){
+  char opt[80];
+  arg=one_argument(arg, opt, cElements(opt));
+  tweakEntry *tweak = tweakInfo[t];
+  TDatabase sneezy(DB_SNEEZY);
+
+  //only one arguements given after type
+  if (!*arg) { 
+    if (is_abbrev(opt, "up")) {
+      tweak->tvalue += .05;
+      tweak->rate = fabs((tweak->tvalue - tweak->cvalue)/30);
+      b->sendTo(format("You tweaked %s up a tad.\r\n") % getTweakName(t));
+      sneezy.query("insert into globaltweaks (tweak_type, tweak_value, tweak_target, tweak_rate) values (%i,%f,%f,%f)", 
```

Comment:
Why on earth do we want history on these?

Reply:
If someone makes a tweakable for xp or something really important one might want to know...